hbase-common/src/main/java/org/apache/hadoop/hbase/io/util/BlockIOUtils.java

```diff
@@ -207,6 +234,17 @@ public static boolean readWithExtra(ByteBuff buf, FSDataInputStream dis, int nec
    */
   public static boolean preadWithExtra(ByteBuff buff, FSDataInputStream dis, long position,
       int necessaryLen, int extraLen) throws IOException {
+    boolean preadbytebuffer = dis.hasCapability("in:preadbytebuffer");
```

Comment:
We need to test it every time? Is it possible to test once when the FSDataInputStream is created?

Reply:
I suppose the overhead is relatively low. I had thought about moving to where the input stream is created, but it would require pretty extensive refactoring.

Internally, hasCapability() uses switch statement to compare the capability string and it should be quite efficient.

I can explore this suggestion later.